### PR TITLE
Add support for making lib features internal

### LINKED ISSUE #115597: `core_intrinsics` should be marked as an internal feature
Using the `intrinsics` feature triggers a warning about this being "internal to the compiler or standard library", but `core_intrinsics` does not. Clearly `core_intrinsics` is internal as well, so it should also get the warning.

`core_intrinsics` is a library feature though, not a language feature -- does the lint support that?

Cc @Nilstrieb

### PR DESCRIPTION
**Pull request by RalfJung**
We have the notion of an "internal" lang feature: a feature that is never intended to be stabilized, and using which can cause ICEs and other issues without that being considered a bug.

This extends that idea to lib features as well. It is an alternative to https://github.com/rust-lang/rust/pull/115623: instead of using an attribute to declare lib features internal, we simply do this based on the name. Everything ending in `_internals` or `_internal` is considered internal.

Then we rename `core_intrinsics` to `core_intrinsics_internal`, which fixes https://github.com/rust-lang/rust/issues/115597.

### CODE REVIEW

**Comment by RalfJung** [comment]
@Amanieu it looks like stdarch is using the core_intrinsics feature, so the build will fail here. What is the best way to change this feature name?

(Submodules are *so painful*. I hope we can transition everything to subtrees eventually...)

File: compiler/rustc_feature/src/unstable.rs [hunk 1]
```diff
@@ -207,9 +216,6 @@ declare_features! (
     (internal, test_2018_feature, "1.31.0", None, Some(Edition::Edition2018)),
     /// Added for testing unstable lints; perma-unstable.
     (internal, test_unstable_lint, "1.60.0", None, None),
-    /// Allows non-`unsafe` —and thus, unsound— access to `Pin` constructions.
-    /// Marked `internal` since perma-unstable and unsound.
-    (internal, unsafe_pin_internals, "1.60.0", None, None),
```
Comment (RalfJung) [inline compiler/rustc_feature/src/unstable.rs:212]: This is a libs feature that was added here solely to make it "incomplete" (https://github.com/rust-lang/rust/commit/c93968aee896ff2e2ccba1b195e778eb32d6d6e3), and then later "internal". That's not needed any more.

**Comment by RalfJung** [comment]
Turns out it's not very hard to avoid using this feature in stdarch, so https://github.com/rust-lang/stdarch/pull/1500 should let us do this transition.

**Comment by RalfJung** [comment]
Oh, turns out the build failure here is from compiler-builtins, not stdarch. stdarch is fine since it is not a separate crate.

**Comment by RalfJung** [comment]
As [suggested](https://github.com/rust-lang/compiler-builtins/pull/557#issuecomment-1820910468) by @Nilstrieb, instead of trying to rename `core_intrinsics` I just added that to the list of internal features now. It's a hack, but it avoids the hassle of having to rename this feature, which is used all over the low-level parts of the standard library.

**Comment by RalfJung** [comment]
`core_panic` looks like another feature that should be covered by this. If the PR gets accepted, we should probably rename it to `core_panic_internals`.

**Comment by RalfJung** [comment]
2 weeks without hearing from the reviewer, let me re-roll the dice.
r? compiler

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
Thanks @RalfJung, this is great

@bors r+

**Pull request merged**